snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,135 @@
+                "timestamp": ClickhouseUnguardedExpression(
+                    f"toStartOfInterval(toDateTime('{timestamp.isoformat()}'), toIntervalSecond({granularity}))"
+                ),
+                "tags.key": ClickhouseNumArray(keys),
+                "tags.value": ClickhouseNumArray(values),
+                **self._process_values(message),
+                "retention_days": ClickhouseInt(message["retention_days"]),
+                "partition": metadata.partition,
+                "offset": metadata.offset,
+                "granularity": ClickhouseInt(granularity),
+            }
+            for granularity in self.GRANULARITIES_SECONDS
+        ]
+        return AggregateInsertBatch(processed, None)
+
+
+class SetsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "s"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        values = message["value"]
+        for v in values:
+            assert isinstance(v, int), "Illegal value in set. Int expected: {v}"
+        value_array = "[" + ",".join([str(v) for v in values]) + "]"
+
+        return {
+            "value": ClickhouseUnguardedExpression(
+                f"arrayReduce('uniqState', {value_array})"
+            )
+        }
+
+
+class CounterAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "c"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        value = message["value"]
+        assert isinstance(
+            value, (int, float)
+        ), "Illegal value for counter value. Int/Float expected {value}"
+
+        return {
+            "value": ClickhouseUnguardedExpression(
+                f"arrayReduce('sumState', [{value}])"
+            )
+        }
+
+
+class DistributionsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "d"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        values = message["value"]
+        for v in values:
+            assert isinstance(
+                v, (int, float)
+            ), "Illegal value in set. Int/Float expected: {v}"
+
+        value_array = "[" + ",".join([str(v) for v in values]) + "]"
+        return {
+            "percentiles": ClickhouseUnguardedExpression(
+                f"arrayReduce('quantilesState(0.5,0.75,0.9,0.95,0.99)', {value_array})"
+            ),
+            "min": ClickhouseUnguardedExpression(
+                f"arrayReduce('minState', {value_array})"
+            ),
+            "max": ClickhouseUnguardedExpression(
+                f"arrayReduce('maxState', {value_array})"
+            ),
```

Comment:
min, max, count and sum may not need to be calculated with the state function. You may just be able to do the math in the consumer and reduce the size of the batch. Though I am not sure, it would be worth trying out.

Reply:
Made this change in my latest commits